tools/smiquery.1
.\"
.\" $Id: smiquery.1.in,v 1.7 2002/11/12 10:03:42 schoenw Exp $
.\"
.TH smiquery 1  "February 11, 2000" "IBR" "SMI Tools"
.SH NAME
smiquery \- query single information from SMI MIB modules
.SH SYNOPSIS
.B smiquery
[
.B "-Vh"
] [
.BI "-p " module
]
.I command name
.SH DESCRIPTION
The \fBsmiquery\fP program is used to query information on a single
item from an MIB module, e.g. module meta information or a single type
definition.
.SH OPTIONS
.TP
\fB-V, --version\fP
Show the smiquery version and exit.
.TP
\fB-h, --help\fP
Show a help text and exit. The help text lists all available \fBcommand\fPs.
.TP
\fB-p \fImodule\fB, --preload \fImodule\fP
Preload the module \fImodule\fP before reading the main module. This may
be helpful if an incomplete main module misses to import some definitions.
.TP
.I command name
This specifies the kind of information to retrieve (module, imports,
revisions, node, parent, compliance, index, members, children, type, or
macro) and the item to retrieve.
.PP
The module in question is searched along a path, which is initialized by
the \fBSMIPATH\fP environment variable or by /usr/local/share/mibs if
\fBSMIPATH\fP is not set. Files with a \fB.sming\fP extension are
expected to be SMIng module files. Files with a \fB.my\fP or without
any extension are expected to be SMIv1/v2 module files.
.SH "EXAMPLE"
This example queries for information on the node ifNumber defined
in the module IF-MIB.
.nf

  $ smiquery node IF-MIB::ifNumber
       MibNode: IF-MIB::ifNumber
	   OID: 1.3.6.1.2.1.2.1
	  Type: Integer32
   Declaration: OBJECT-TYPE
      NodeKind: scalar
	Access: read-only
	Status: current
   Description: The number of network interfaces (regardless of their
		current state) present on this system.

.fi
.SH "SEE ALSO"
The
.BR libsmi (3)
project is documented at
.BR "http://www.ibr.cs.tu-bs.de/projects/libsmi/" "."
.SH "AUTHOR"
(C) 1999, 2000 Frank Strauss, TU Braunschweig, Germany <strauss@ibr.cs.tu-bs.de>
.br


ChangeLog

2008-02-27  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* tools/Makefile.am: added missing files to the distribution.
	* libsmi-0.4.7.tar.gz: released version 0.4.7.

2008-02-27  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/URI-TC-MIB: added (RFC 5017).
	* mibs/ietf/PIM-STD-MIB: added (RFC 5060).
	* mibs/ietf/IF-CAP-STACK-MIB,EFM-CU-MIB: added (RFC 5066).
	* mibs/ietf/UDPLITE-MIB: added (RFC 5097).
	* mibs/ietf/LANGTAG-TC-MIB: added (RFC 5131).
	* mibs/ietf/IPMCAST-MIB: added (RFC 5132).
	* updated IANA MIBs.
	* libsmi-0.4.6.tar.gz: released version 0.4.6.

2007-09-03  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/T11-FC-RSCN-MIB: added (RFC 4983).

2007-08-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/T11-FC-FABRIC-CONFIG-SERVER-MIB: added (RFC 4935).
	* mibs/ietf/T11-FC-FABRIC-LOCK-MIB,T11-FC-ZONE-SERVER-MIB: added
	  (RFC 4936).
	* mibs/ietf/ISNS-MIB: added (RFC 4939).

2007-07-19  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DOT3-EPON-MIB: added (RFC 4837).
	* mibs/ietf/DOT3-OAM-MIB: added (RFC 4878).

2007-06-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/TCP-ESTATS-MIB: added (RFC 4898).

2007-05-08  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/SIP-{UA,TC,COMMON,SERVER}-MIB: added (RFC 4780).
	* mibs/ietf/MAU-MIB: updated (RFC 4836).

2007-03-05  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DS1-MIB: updated (RFC 4805).
	* mibs/ietf/IPSEC-SPD-MIB: added (RFC 4807).

2007-03-05  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/iana/IANA-GMPLS-TC-MIB,IANA-IPPM-METRICS-REGISTRY-MIB,
	  IANA-ITU-ALARM-TC-MIB,IANA-MAU-MIB: added.
	* mibs/iana/IANA-CHARSET-MIB,IANA-ADDRESS-FAMILY-NUMBERS-MIB,
	  IANAifType-MIB: added.
	* mibs/ietf/GMPLS-TC-STD-MIB (RFC 4801): added.
	* mibs/ietf/GMPLS-TE-STD-MIB (RFC 4802): added.
	* mibs/ietf/GMPLS-LSR-STD-MIB,GMPLS-LABEL-STD-MIB (RFC 4803): added.

2007-02-01  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/MSDP-MIB: added (RFC 4624).
	* mibs/ietf/PKTC-IETF-MTA-MIB: added (RFC 4682).
	* mibs/ietf/ADSL2-LINE-MIB,ADSL2-LINE-TC-MIB: added (RFC 4706).
	* mibs/ietf/RAQMON-MIB: added (RFC 4711).
	* mibs/ietf/T11-FC-VIRTUAL-FABRIC-MIB: added (RFC 4747).
	* mibs/ietf/DOCS-CABLE-DEVICE-MIB: updated (RFC 4639).
	* mibs/ietf/OSPF-MIB,OSPF-TRAP-MIB: updated (RFC 4750).
	* lib/scanner-smi.l: accept also \r as end-of-line.

2006-10-04  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/T11-FC-ROUTE-MIB: added (RFC 4625).
	* mibs/ietf/T11-FC-FSPF-MIB: added (RFC 4626).
	* mibs/ietf/LMP-MIB: updated (RFC 4631).
	* mibs/ietf/RADIUS-DYNAUTH-CLIENT-MIB: added (RFC 4672).
	* mibs/ietf/RADIUS-DYNAUTH-SERVER-MIB: added (RFC 4673).

2006-09-06  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* autogen.sh: Use glibtool instead of libtool on Darwin.
	Thanks to Sebastian.

2006-08-30  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/RADIUS-AUTH-CLIENT-MIB: updated (RFC 4668).
	* mibs/ietf/RADIUS-AUTH-SERVER-MIB: updated (RFC 4669).
	* mibs/ietf/RADIUS-ACC-CLIENT-MIB: updated (RFC 4670).
	* mibs/ietf/RADIUS-ACC-SERVER-MIB: updated (RFC 4671).

2006-07-15  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DOCS-IF-MIB: updated (RFC 4546).
	* mibs/ietf/DOCS-IETF-CABLE-DEVICE-NOTIFICATION-MIB: added (RFC 4547).
	* mibs/ietf/DISMAN-NSLOOKUP-MIB,DISMAN-PING-MIB,DISMAN-TRACEROUTE-MIB:
	  updated (RFC 4560).

2006-06-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/ISCSI-MIB: added (RFC 4544).
	* mibs/ietf/IPS-AUTH-MIB: added (RFC 4545).

2006-05-30  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/SNMP-USM-AES-MIB: updated (RFC 3826).

2006-05-29  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/BGP4-MIB: updated (RFC 4273).
	* mibs/ietf/IP-FORWARD-MIB: updated (RFC 4292).
	* mibs/ietf/IP-MIB: updated (RFC 4293).
	* mibs/ietf/MOBILEIPV6-MIB: added (RFC 4295).
	* mibs/ietf/RSTP-MIB: added (RFC 4318).
	* mibs/ietf/DOCS-IETF-QOS-MIB: added (RFC 4323).
	* mibs/ietf/LMP-MIB: added (RFC 4327).
	* mibs/ietf/{P-BRIDGE-MIB,Q-BRIDGE-MIB}: updated (RFC 4363).
	* mibs/ietf/{MPLS-LC-ATM-STD-MIB,MPLS-LC-FR-STD-MIB):
	  added (RFC 4368).
	* mibs/ietf/IFCP-MGMT-MIB: added (RFC 4369).
	* mibs/ietf/MPLS-L3VPN-STD-MIB: added (RFC 4382).
	* mibs/ietf/FCIP-MGMT-MIB: added (RFC 4404).
	* mibs/ietf/T11-FC-NAME-SERVER-MIB: added (RFC 4438).
	* mibs/ietf/{T11-TC-MIB,T11-FC-FABRIC-ADDR-MGR-MIB}:
	  added (RFC 4439).
	* mibs/ietf/ISIS-MIB: added (RFC 4444).
	* mibs/ietf/SCSI-MIB: added (RFC 4455).
	* mibs/ietf/{AGGREGATE-MIB,TIME-AGGREGATE-MIB}: added (RFC 4498).
	* mibs/ietf/RMON2-MIB: updated (RFC 4502).

	* libsmi-0.4.5.tar.gz: released version 0.4.5.
	There was a bug that breaks compliation of 0.4.4, so here
	is a new release.

2006-05-16  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* libsmi-0.4.4.tar.gz: released version 0.4.4.
	This is more or less just a snapshot to give package
	maintainers a fresh and handy release number.

2005-12-18  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/HDSL2-SHDSL-LINE-MIB: updated (RFC 4319).

2005-12-16  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/{ENTITY-STATE-MIB,ENTITY-STATE-TC-MIB}:
	  added (RFC 4268).

2005-11-25  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/TE-LINK-STD-MIB: added (RFC 4220).

	* mibs/ietf/VPN-TC-STD-MIB: added (RFC 4265).

2005-10-10  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/DOCS-IETF-BPI2-MIB: added (RFC 4131).

	* mibs/ietf/ENTITY-MIB: updated (RFC 4133).

	* mibs/ietf/BRIDGE-MIB: updated (RFC 4188).

2005-10-10  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/SSPM-MIB: added (RFC 4149).

2005-06-23  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/TUNNEL-MIB: updated (RFC 4087).

	* mibs/ietf/UDP-MIB: updated (RFC 4113).

2005-05-17  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/FC-MGMT-MIB: added (RFC 4044).

2005-05-12  Frank Strauss  <strauss@ibr.cs.tu-bs.de>

	* mibs/ietf/VDSL-LINE-EXT-SCM-MIB: added (RFC 4069).

	* mibs/ietf/VDSL-LINE-EXT-MCM-MIB: added (RFC 4070).


tools/smiquery.c
/*
 * smiquery.c --
 *
 *      A simple SMI query program, using libsmi.
 *
 * Copyright (c) 1999 Frank Strauss, Technical University of Braunschweig.
 *
 * See the file "COPYING" for information on usage and redistribution
 * of this file, and for a DISCLAIMER OF ALL WARRANTIES.
 *
 * @(#) $Id: smiquery.c,v 1.72 2002/12/06 15:37:46 strauss Exp $
 */

#include <config.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif
#ifdef HAVE_WIN_H
#include "win.h"
#endif

#include "smi.h"
#include "shhopt.h"



static char *smiStringStatus(SmiStatus status)
{
    return
	(status == SMI_STATUS_CURRENT)     ? "current" :
	(status == SMI_STATUS_DEPRECATED)  ? "deprecated" :
	(status == SMI_STATUS_OBSOLETE)    ? "obsolete" :
	(status == SMI_STATUS_MANDATORY)   ? "mandatory" :
	(status == SMI_STATUS_OPTIONAL)    ? "optional" :
					     "<UNDEFINED>";
}

static char *smiStringAccess(SmiAccess access)
{
    return
	(access == SMI_ACCESS_NOT_ACCESSIBLE) ? "not-accessible" :
	(access == SMI_ACCESS_NOTIFY)	      ? "accessible-for-notify" :
	(access == SMI_ACCESS_READ_ONLY)      ? "read-only" :
	(access == SMI_ACCESS_READ_WRITE)     ? "read-write" :
						"<UNDEFINED>";
}

static char *smiStringLanguage(SmiLanguage language)
{
    return
	(language == SMI_LANGUAGE_UNKNOWN)    ? "<unknown>" :
	(language == SMI_LANGUAGE_SMIV1)      ? "SMIv1" :
	(language == SMI_LANGUAGE_SMIV2)      ? "SMIv2" :
	(language == SMI_LANGUAGE_SMING)      ? "SMIng" :
						"<UNDEFINED>";
}

static char *smiStringDecl(SmiDecl macro)
{
    return
        (macro == SMI_DECL_UNKNOWN)           ? "<unknown>" :
        (macro == SMI_DECL_IMPLICIT_TYPE)     ? "<implicit>" :
        (macro == SMI_DECL_TYPEASSIGNMENT)    ? "<type-assignment>" :
        (macro == SMI_DECL_IMPL_SEQUENCEOF)   ? "<implicit-sequence-of>" :
        (macro == SMI_DECL_VALUEASSIGNMENT)   ? "<value-assignment>" :
        (macro == SMI_DECL_OBJECTTYPE)        ? "OBJECT-TYPE" :
        (macro == SMI_DECL_OBJECTIDENTITY)    ? "OBJECT-IDENTITY" :
        (macro == SMI_DECL_MODULEIDENTITY)    ? "MODULE-IDENTITY" :
        (macro == SMI_DECL_NOTIFICATIONTYPE)  ? "NOTIFICATIONTYPE" :
        (macro == SMI_DECL_TRAPTYPE)          ? "TRAP-TYPE" :
        (macro == SMI_DECL_OBJECTGROUP)       ? "OBJECT-GROUP" :
        (macro == SMI_DECL_NOTIFICATIONGROUP) ? "NOTIFICATION-GROUP" :
        (macro == SMI_DECL_MODULECOMPLIANCE)  ? "MODULE-COMPLIANCE" :
        (macro == SMI_DECL_AGENTCAPABILITIES) ? "AGENT-CAPABILITIES" :
        (macro == SMI_DECL_TEXTUALCONVENTION) ? "TEXTUAL-CONVENTION" :
        (macro == SMI_DECL_MODULE)	      ? "module" :
        (macro == SMI_DECL_TYPEDEF)	      ? "typedef" :
        (macro == SMI_DECL_NODE)	      ? "node" :
        (macro == SMI_DECL_SCALAR)	      ? "scalar" :
        (macro == SMI_DECL_TABLE)	      ? "table" :
        (macro == SMI_DECL_ROW)		      ? "row" :
        (macro == SMI_DECL_COLUMN)	      ? "column" :
        (macro == SMI_DECL_NOTIFICATION)      ? "notification" :
        (macro == SMI_DECL_GROUP)	      ? "group" :
        (macro == SMI_DECL_COMPLIANCE)	      ? "compliance" :
                                                "<UNDEFINED>";
}

static char *smiStringNodekind(SmiNodekind nodekind)
{
    return
        (nodekind == SMI_NODEKIND_UNKNOWN)      ? "<unknown>" :
        (nodekind == SMI_NODEKIND_NODE)         ? "node" :
        (nodekind == SMI_NODEKIND_SCALAR)       ? "scalar" :
        (nodekind == SMI_NODEKIND_TABLE)        ? "table" :
        (nodekind == SMI_NODEKIND_ROW)          ? "row" :
        (nodekind == SMI_NODEKIND_COLUMN)       ? "column" :
        (nodekind == SMI_NODEKIND_NOTIFICATION) ? "notification" :
        (nodekind == SMI_NODEKIND_GROUP)        ? "group" :
        (nodekind == SMI_NODEKIND_COMPLIANCE)   ? "compliance" :
        (nodekind == SMI_NODEKIND_CAPABILITIES) ? "capabilities" :
                                                  "<UNDEFINED>";
}

static char *smiStringBasetype(SmiBasetype basetype)
{
    return
        (basetype == SMI_BASETYPE_UNKNOWN)           ? "<unknown>" :
        (basetype == SMI_BASETYPE_OCTETSTRING)       ? "OctetString" :
        (basetype == SMI_BASETYPE_OBJECTIDENTIFIER)  ? "ObjectIdentifier" :
        (basetype == SMI_BASETYPE_UNSIGNED32)        ? "Unsigned32" :
        (basetype == SMI_BASETYPE_INTEGER32)         ? "Integer32" :
        (basetype == SMI_BASETYPE_UNSIGNED64)        ? "Unsigned64" :
        (basetype == SMI_BASETYPE_INTEGER64)         ? "Integer64" :
        (basetype == SMI_BASETYPE_FLOAT32)           ? "Float32" :
        (basetype == SMI_BASETYPE_FLOAT64)           ? "Float64" :
        (basetype == SMI_BASETYPE_FLOAT128)          ? "Float128" :
        (basetype == SMI_BASETYPE_ENUM)              ? "Enumeration" :
        (basetype == SMI_BASETYPE_BITS)              ? "Bits" :
                                                   "<UNDEFINED>";
}



static char *format(const char *s)
{
    static char ss[20000];
    int i, j;

    if (!s) {
	sprintf(ss, "-");
    } else {
	for(i = 0, j = 0; s[i]; i++) {
	    ss[j++] = s[i];
	    if (s[i] == '\n') {
		sprintf(&ss[j], "              ");
		j += 14;
	    }
	}
	ss[j] = 0;
    }
    return ss;
}


static void usage()
{
    fprintf(stderr,
	    "Usage: smiquery [options] command name [command name ...]\n"
	    "  -V, --version        show version and license information\n"
	    "  -h, --help           show usage information\n"
	    "  -c, --config=file    load a specific configuration file\n"
	    "  -p, --preload=module preload <module>\n"
	    "\nSupported commands are:\n"
	    "  module <module>           show information on module <module>\n"
	    "  imports <module>          show import list of module <module>\n"
	    "  node <module::name>       show information on node <module::name>\n"
	    "  compliance <module::name> show information on compliance node <module::name>\n"
	    "  children <module::name>   show children list of node <module::name>\n"
	    "  type <module::name>       show information on type <module::name>\n"
	    "  macro <module::name>      show information on macro <module::name>\n");
}



static void help() { usage(); exit(0); }
static void version() { printf("smiquery " SMI_VERSION_STRING "\n"); exit(0); }
static void config(char *filename) { smiReadConfig(filename, "smiquery"); }
static void preload(char *module) { smiLoadModule(module); }



int main(int argc, char *argv[])
{
    SmiModule *module;
    SmiNode *node, *node2, *child, *parent;
    SmiType *type, *parenttype;
    SmiMacro *macro;
    SmiNamedNumber *nn;
    SmiRange *range;
    SmiImport *import;
    SmiRevision *revision;
    SmiOption *option;
    SmiRefinement *refinement;
    SmiElement *element;
    char *command, *name, *p;
    int flags, i;
    char s1[40], s2[40];

    static optStruct opt[] = {
	/* short long              type        var/func       special       */
	{ 'h', "help",           OPT_FLAG,   help,          OPT_CALLFUNC },
	{ 'V', "version",        OPT_FLAG,   version,       OPT_CALLFUNC },
	{ 'c', "config",         OPT_STRING, config,        OPT_CALLFUNC },
	{ 'p', "preload",        OPT_STRING, preload,       OPT_CALLFUNC },
	{ 0, 0, OPT_END, 0, 0 }  /* no more options */
    };
    
    for (i = 1; i < argc; i++)
	if ((strstr(argv[i], "-c") == argv[i]) ||
	    (strstr(argv[i], "--config") == argv[i])) break;
    if (i == argc) 
	smiInit("smiquery");
    else
	smiInit(NULL);

    flags = smiGetFlags();

    optParseOptions(&argc, argv, opt, 0);

    if (!(argc % 2)) {
	usage();
	return 1;
    }

    for (i = 1; i < argc; i +=2) {

	command = argv[i];
	name = argv[i+1];

	if (!strcmp(command, "module")) {
	    module = smiGetModule(name);
	    if (module) {
		node = smiGetModuleIdentityNode(module);
		printf("      Module: %s\n", format(module->name));
		if (module->path)
		    printf("    Pathname: %s\n", module->path);
		if (node)
		    printf("      Object: %s\n",
			   smiRenderNode(node, SMI_RENDER_ALL));
		if (module->organization)
		    printf("Organization: %s\n", format(module->organization));
		if (module->contactinfo)
		    printf(" ContactInfo: %s\n", format(module->contactinfo));
		if (module->description)
		    printf(" Description: %s\n", format(module->description));
		if (module->reference)
		    printf("   Reference: %s\n", format(module->reference));
		printf("    Language: %s\n", smiStringLanguage(module->language));
		printf(" Conformance: %d\n", module->conformance);
		printf("      Loaded: %s\n", smiIsLoaded(name) ? "yes" : "no");

		for(revision = smiGetFirstRevision(module);
		    revision ; revision = smiGetNextRevision(revision)) {
		    printf("    Revision: %s", ctime(&revision->date));
		    if (revision->description)
			printf(" Description: %s\n", format(revision->description));
		}
	    }
	} else if (!strcmp(command, "imports")) {
	    module = smiGetModule(name);
	    if (module && smiGetFirstImport(module)) {
		printf("     Imports:");
		for(import = smiGetFirstImport(module); import ; ) {
		    printf(" %s::%s", import->module, import->name);
		    import = smiGetNextImport(import);
		    if (import) {
			printf("\n             ");
		    }
		}
		printf("\n");
	    }
	} else if (!strcmp(command, "node")) {
	    node = smiGetNode(NULL, name);
	    if (node) {
		parent = smiGetParentNode(node);
	    }
	    if (node) {
		type = smiGetNodeType(node);
		printf("     MibNode: %s\n", smiRenderNode(node, SMI_RENDER_ALL));
		printf("         OID: %s\n", smiRenderOID(node->oidlen, node->oid,
							  0));
		if (parent)
		    printf("  ParentNode: %s\n",
			   smiRenderNode(parent, SMI_RENDER_ALL));
		if (type)
		    printf("        Type: %s\n",
			   smiRenderType(type, SMI_RENDER_ALL));
		if (node->value.basetype != SMI_BASETYPE_UNKNOWN)
		    printf("     Default: %s\n", smiRenderValue(&node->value, type,
								SMI_RENDER_ALL));
		if (node->decl != SMI_DECL_UNKNOWN)
		    printf(" Declaration: %s\n", smiStringDecl(node->decl));
		printf("    NodeKind: %s\n", smiStringNodekind(node->nodekind));
		if (node->nodekind == SMI_NODEKIND_ROW) {
		    printf ("   Creatable: %s\n", node->create ? "yes" : "no");
		    printf ("     Implied: %s\n", node->implied ? "yes" : "no");
		}
		switch (node->nodekind) {
		case SMI_NODEKIND_ROW: p = "Index"; break;
		case SMI_NODEKIND_COMPLIANCE: p = "Mandatory"; break;
		case SMI_NODEKIND_CAPABILITIES: p = "Includes"; break;
		case SMI_NODEKIND_GROUP: p = "Members"; break;
		case SMI_NODEKIND_NOTIFICATION: p = "Objects"; break;
		default: p = "Elements";
		}
		if (smiGetFirstElement(node)) {
		    printf("%12s:", p);
		    for(element = smiGetFirstElement(node);
			element ; ) {
			node2 = smiGetElementNode(element);
			printf(" %s", smiRenderNode(node2, SMI_RENDER_ALL));
			element = smiGetNextElement(element);
			if (element) {
			    printf("\n             ");
			}
		    }
		    printf("\n");
		}
		if (node->access != SMI_ACCESS_UNKNOWN)
		    printf("      Access: %s\n", smiStringAccess(node->access));
		if (node->status != SMI_STATUS_UNKNOWN)
		    printf("      Status: %s\n", smiStringStatus(node->status));
		if (node->format)
		    printf("      Format: %s\n", format(node->format));
		if (node->units)
		    printf("       Units: %s\n", format(node->units));
		if (node->description)
		    printf(" Description: %s\n", format(node->description));
		if (node->reference)
		    printf("   Reference: %s\n", format(node->reference));
	    }
	} else if (!strcmp(command, "compliance")) {
	    node = smiGetNode(NULL, name);
	    if (node) {
		if (smiGetFirstElement(node)) {
		    printf("   Mandatory:");
		    for(element = smiGetFirstElement(node);
			element ; ) {
			node2 = smiGetElementNode(element);
			printf(" %s", smiRenderNode(node2, SMI_RENDER_ALL));
			element = smiGetNextElement(element);
			if (element) {
			    printf("\n             ");
			}
		    }
		    printf("\n");
		}
		if (smiGetFirstOption(node)) {
		    for(option = smiGetFirstOption(node); option ;) {
			node2 = smiGetOptionNode(option);
			printf("      Option: %s\n",
			       smiRenderNode(node2, SMI_RENDER_ALL));
			if (option->description)
			    printf(" Description: %s\n",
				   format(option->description));
			option = smiGetNextOption(option);
		    }
		}
		if (smiGetFirstRefinement(node)) {
		    for(refinement = smiGetFirstRefinement(node);
			refinement ;
			refinement = smiGetNextRefinement(refinement)) {
			node2 = smiGetRefinementNode(refinement);
			printf("  Refinement: %s\n",
			       smiRenderNode(node2, SMI_RENDER_ALL));
			type = smiGetRefinementType(refinement);
			if (type) {
			    printf("        Type: %s\n",
				   smiRenderType(type, SMI_RENDER_ALL));
			}
			type = smiGetRefinementWriteType(refinement);
			if (type) {
			    module = smiGetTypeModule(type);
			    printf("  Write-Type: %s\n",
				   smiRenderType(type, SMI_RENDER_ALL));
			}
			if (refinement->access != SMI_ACCESS_UNKNOWN) {
			    printf("      Access: %s\n",
				   smiStringAccess(refinement->access));
			}
			if (refinement->description)
			    printf(" Description: %s\n",
				   format(refinement->description));
		    }
		}
	    }
	} else if (!strcmp(command, "children")) {
	    node = smiGetNode(NULL, name);
	    if (node && smiGetFirstChildNode(node)) {
		printf("    Children:");
		for(child = smiGetFirstChildNode(node);
		    child ; ) {
		    printf(" %s", smiRenderNode(child, SMI_RENDER_ALL));
		    child = smiGetNextChildNode(child);
		    if (child) {
			printf("\n             ");
		    }
		}
		printf("\n");
	    }
	} else if (!strcmp(command, "type")) {
	    p = strrchr(name, ':');
	    if (!p) p = strrchr(name, '.');
	    if (!p) p = strrchr(name, '!');
	    if (p) {
		p++;
	    } else {
		p = name;
	    }
	    if (islower((int)name[0]) || isdigit((int)name[0]) ||
		!isupper((int)p[0])) {
		node = smiGetNode(NULL, name);
		if (node) {
		    type = smiGetNodeType(node);
		} else {
		    type = smiGetType(NULL, name);
		}
	    } else {
		type = smiGetType(NULL, name);
	    }
	    if (type) {
		parenttype = smiGetParentType(type);
		printf("        Type: %s\n", smiRenderType(type, SMI_RENDER_ALL));
		printf("    Basetype: %s\n", smiStringBasetype(type->basetype));
		if (parenttype)
		    printf(" Parent Type: %s\n",
			   smiRenderType(parenttype, SMI_RENDER_ALL));
		if (type->value.basetype != SMI_BASETYPE_UNKNOWN)
		    printf("     Default: %s\n", smiRenderValue(&type->value, type,
								SMI_RENDER_ALL));
		if ((type->basetype == SMI_BASETYPE_ENUM) ||
		    (type->basetype == SMI_BASETYPE_BITS)) {
		    if (smiGetFirstNamedNumber(type)) {
			printf("     Numbers:");
			for(nn = smiGetFirstNamedNumber(type);
			    nn ; nn = smiGetNextNamedNumber(nn)) {
			    printf(" %s(%ld)",
				   nn->name, nn->value.value.integer32);
			}
			printf("\n");
		    }
		} else {
		    if (smiGetFirstRange(type)) {
			printf("      Ranges:");
			for(range = smiGetFirstRange(type);
			    range ; range = smiGetNextRange(range)) {
			    strcpy(s1, smiRenderValue(&range->minValue, type, 0));
			    strcpy(s2, smiRenderValue(&range->maxValue, type, 0));
			    printf(" %s", s1);
			    if (strcmp(s1, s2)) printf("..%s", s2);
			}
			printf("\n");
		    }
		}
		printf(" Declaration: %s\n", smiStringDecl(type->decl));
		if (type->status != SMI_STATUS_UNKNOWN)
		    printf("      Status: %s\n", smiStringStatus(type->status));
		if (type->format)
		    printf("      Format: %s\n", format(type->format));
		if (type->units)
		    printf("       Units: %s\n", format(type->units));
		if (type->description)
		    printf(" Description: %s\n", format(type->description));
		if (type->reference)
		    printf("   Reference: %s\n", format(type->reference));
	    }
	} else if (!strcmp(command, "macro")) {
	    macro = smiGetMacro(NULL, name);
	    if (macro) {
		printf("       Macro: %s\n", format(macro->name));
		if (macro->status != SMI_STATUS_UNKNOWN)
		    printf("      Status: %s\n", smiStringStatus(macro->status));
		if (macro->description)
		    printf(" Description: %s\n", format(macro->description));
		if (macro->reference)
		    printf("   Reference: %s\n", format(macro->reference));
	    }
	} else {
	    usage();
	    return 1;
	}
    }
    
    smiExit();
    
    return 0;
}
